lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp

```diff
@@ -16,18 +16,332 @@
 #include "circt/Dialect/FIRRTL/FIRRTLOps.h"
 #include "circt/Dialect/FIRRTL/FIRRTLTypes.h"
 #include "circt/Dialect/FIRRTL/FIRRTLVisitors.h"
+#include "circt/Dialect/FIRRTL/Namespace.h"
 #include "circt/Dialect/FIRRTL/Passes.h"
+#include "circt/Dialect/HW/HWAttributes.h"
 #include "circt/Support/LLVM.h"
 #include "mlir/IR/BlockAndValueMapping.h"
+#include "llvm/ADT/BitVector.h"
+#include "llvm/ADT/SetOperations.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/FormatVariadic.h"
+
+#define DEBUG_TYPE "firrtl-inliner"
 
 using namespace circt;
 using namespace firrtl;
 using namespace chirrtl;
 
+using llvm::BitVector;
+
 //===----------------------------------------------------------------------===//
 // Module Inlining Support
 //===----------------------------------------------------------------------===//
 
+namespace {
+/// A representation of an NLA that can be mutated.  This is intended to be used
+/// in situations where you want to make a series of modifications to an NLA
+/// while also being able to query information about it.  Finally, the NLA is
+/// written back to the IR to replace the original NLA.
+class MutableNLA {
+  // Storage of the NLA this represents.
+  NonLocalAnchor nla;
+
+  // A namespace that can be used to generate new symbol names if needed.
+  CircuitNamespace *circuitNamespace;
+
+  /// A mapping of symbol to index in the NLA.
+  DenseMap<Attribute, unsigned> symIdx;
+
+  /// Records which elements of the path are inlined.
+  BitVector inlinedSymbols;
+
+  /// The point after which the NLA is flattened.  A value of "-1" indicates
+  /// that this was never set.
+  signed flattenPoint = -1;
+
+  /// Indicates if the _original_ NLA is dead and should be deleted.  Updates
+  /// may still need to be written if the newTops vector below is non-empty.
+  bool dead = false;
+
+  /// Stores new roots for the NLA.  If this is non-empty, then it indicates
+  /// that the NLA should be copied and re-topped using the roots stored here.
+  /// This is non-empty when athe NLA's root is inlined and the original NLA
+  /// migrates to each instantiator of the original NLA.
+  SmallVector<std::pair<StringAttr, StringAttr>> newTops;
+
+  /// Cache of roots that this module participates in.  This is only valid when
+  /// newTops is non-empty.
+  DenseSet<StringAttr> rootSet;
+
+  /// Stores the size of the NLA path.
+  unsigned int size;
+
+public:
+  MutableNLA(NonLocalAnchor nla, CircuitNamespace *circuitNamespace)
+      : nla(nla), circuitNamespace(circuitNamespace),
+        inlinedSymbols(BitVector(nla.namepath().size(), true)),
+        size(nla.namepath().size()) {
+    for (size_t i = 0, e = size; i != e; ++i)
+      symIdx.insert({modPart(i), i});
+  }
+
+  /// This default constructor only exists so that the `[]` referential lookup
+  /// can be used for `Inliner::nlaMap`.  The `nlaMap` is prepopulated with all
+  /// NLAs, so this will always hit.
```

Comment:
Comment is unclear. I can fix it. Default constructor should _never_ be used here and if it is, it's a hard error.

I want to use a `DenseMap<Attribute, MutableNLA> nlaMap` to do a lookup of the NLA symbol to its mutable NLA.  However, use of `DenseMap` mandates a default constructor because `[]` is going to call this if it doesn't get a hit.  I could do default construction of the `MutableNLA`, but it would require examining the IR, e.g., find the parent circuit, find the NLA that matches, then construct.  Instead, I am guaranteeing that I prepopulate the `nlaMap` and I'm providing a default constructor that errors (which is indicative of the pass finding an NLA symbol that doesn't match any existing NLA) so that I can still use `DenseMap`, but not make it thread unsafe by allowing it to do default construction.

There may be a better way of describing this.

Reply:
I clarified this comment after rebasing.